src/ape/managers/chain.py

```diff
@@ -975,71 +980,100 @@ def get(
             # Contract could be a minimal proxy
             proxy_info = self._local_proxies.get(address_key) or self._get_proxy_info_from_disk(
                 address_key
             )
 
+            # Attempt to retrieve missing proxy info
             if not proxy_info:
                 proxy_info = self.provider.network.ecosystem.get_proxy_info(address_key)
-                if proxy_info and self._is_live_network:
+                if proxy_info and self._is_live_network and cache_to_disk:
+                    # If found, cache to disk
                     self._cache_proxy_info_to_disk(address_key, proxy_info)
 
+            # If proxy info is already known, save the proxy info to memory and
+            # call this function again at the implementation contract address
             if proxy_info:
                 self._local_proxies[address_key] = proxy_info
-                return self.get(proxy_info.target, default=default)
+                return self.get(proxy_info.target, default=default, cache_to_disk=cache_to_disk)
 
             if not self.provider.get_code(address_key):
+                # If no code is stored at this address, store the provided default contract
+                # type in memory. This block is skipped if a contract type was not provided
+                # (`default==None`)
                 if default:
                     self._local_contract_types[address_key] = default
-                    self._cache_contract_to_disk(address_key, default)
+                    if cache_to_disk:
+                        self._cache_contract_to_disk(address_key, default)
 
                 return default
 
-            # Also gets cached to disk for faster lookup next time.
-            contract_type = self._get_contract_type_from_explorer(address_key)
+        # Attempt to get the contract type from block explorer
+        # (unless overridden by `fetch_from_explorer=False`)
+        contract_type_from_explorer = (
+            self._get_contract_type_from_explorer(address_key) if fetch_from_explorer else None
+        )
+
+        # Identify the 'canonical' contract type before performing cache updates
+        contract_type = None
+        if default is not None:
+            # If a contract type was provided, use it
+            contract_type = default
+        elif contract_type_from_explorer is not None and contract_type_from_explorer != default:
+            # If a contract type was not provided, and one was found from the block explorer, use it
+            contract_type = contract_type_from_explorer
+        elif contract_type_from_disk is not None and contract_type_from_disk != default:
+            # If a contract type was not provided, and one was found from disk, use it
+            contract_type = contract_type_from_disk
+        elif contract_type_from_memory is not None and contract_type_from_memory != default:
+            contract_type = contract_type_from_memory
+        else:
+            raise ValueError(
```

Comment:
In the `get()` method, when the contract type is not found,  it should return `None`.
The `__getitem__` method is the one that raises.
This mimics the Pythonic dictionary.

Adding this raise statement is a breaking change in that sense as well.

Reply:
Ah, I understand the intent here. I will rework this to eliminate raising the exception.

Reply:
I've removed the exception, falling through all the `if`/`elif` checks will preserve the return behavior since `contract_type=None` if not set to a type from explorer, disk cache, or memory